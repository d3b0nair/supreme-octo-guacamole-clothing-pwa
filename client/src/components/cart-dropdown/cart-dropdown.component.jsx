import React from "react";
import { connect } from "react-redux";
import { selectCartItems } from "../../redux/cart/cart.selector";
import { createStructuredSelector } from "reselect";
import { withRouter } from "react-router-dom";
import { toggleCartHidden } from "../../redux/cart/cart.actions";
import CartItem from "../cart-item/cart-item.component";
import * as handleClickHelper from "./cart-dropdown.component";

import {
  CartDropdownContainer,
  CartItemsContainer,
  ButtonWrapper,
  EmptyMsgWrapper,
} from "./cart-dropdown.styles";

export const handleClick = (history, dispatch) => {
  history.push("/checkout");
  dispatch(toggleCartHidden());
};
export function CartDropdown({ cartItems, history, dispatch }) {
  return (
    <CartDropdownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMsgWrapper>Your cart is empty</EmptyMsgWrapper>
        )}
      </CartItemsContainer>
      <ButtonWrapper
        onClick={() => handleClickHelper.handleClick(history, dispatch)}
      >
        GO TO CHECKOUT
      </ButtonWrapper>
    </CartDropdownContainer>
  );
}
const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

export default withRouter(connect(mapStateToProps)(CartDropdown));
