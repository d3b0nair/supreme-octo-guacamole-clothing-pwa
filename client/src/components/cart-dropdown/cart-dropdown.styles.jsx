import styled from "styled-components";
import CustomButton from "../custom-button/custom-button.component";
import { frostedGlass, customScrollBar } from "./../../global.styles";

export const CartDropdownContainer = styled.div`
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  ${frostedGlass}
`;

export const CartItemsContainer = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  ${customScrollBar}
`;

export const ButtonWrapper = styled(CustomButton)`
  margin-top: auto;
`;
ButtonWrapper.displayName = "ButtonWrapper";

export const EmptyMsgWrapper = styled.span`
  font-size: 18px;
  margin: 50px auto;
`;
