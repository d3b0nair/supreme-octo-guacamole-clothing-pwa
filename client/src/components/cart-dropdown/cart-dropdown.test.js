import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import * as Component from "./cart-dropdown.component";

const mockProps = (otherProps) => ({
  ...otherProps,
  history: { push: jest.fn() },
  cartItems: [],
  dispatch: jest.fn(),
});

const component = (mockProps) => <Component.CartDropdown {...mockProps} />;

afterEach(cleanup);

describe("CartDropdown component", () => {
  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });
  it("on button click it should open checkout page and hide itself", async () => {
    render(component(mockProps()));
    const handleClick = jest.spyOn(Component, "handleClick");
    await waitFor(() => fireEvent.click(screen.getByRole("button")));
    expect(handleClick).toHaveBeenCalled();
  });
});
