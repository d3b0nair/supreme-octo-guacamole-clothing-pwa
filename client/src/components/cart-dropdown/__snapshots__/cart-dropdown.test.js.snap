// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should mount 1`] = `
<div>
  <div
    class="sc-eCImPb ijRBdG"
  >
    <div
      class="sc-jRQBWg fEnDWv"
    >
      <span
        class="sc-iCfMLu bXFgRT"
      >
        Your cart is empty
      </span>
    </div>
    <button
      class="sc-hKwDye hzDzeB sc-gKclnd cjNBCc"
    >
      GO TO CHECKOUT
    </button>
  </div>
</div>
`;
